istioctl/cmd/revision.go

```diff
@@ -0,0 +1,659 @@
+	"istio.io/istio/operator/pkg/util/clog"
+	"istio.io/istio/pkg/config"
+	v1 "k8s.io/api/core/v1"
+	meta_v1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	apimachinery_schema "k8s.io/apimachinery/pkg/runtime/schema"
+	"k8s.io/apimachinery/pkg/util/duration"
+	"k8s.io/apimachinery/pkg/util/validation"
+)
+
+type revisionArgs struct {
+	manifestsPath string
+	name          string
+	verbose       bool
+	sections      []string
+	output        string
+}
+
+const (
+	IstioOperatorCRSection = "cr"
+	ControlPlaneSection    = "cp"
+	GatewaysSection        = "gw"
+	PodsSection            = "po"
+
+	// TODO: This should be moved to istio/api:label package
+	istioTag = "istio.io/tag"
+)
+
+var (
+	validFormats = []string{"table", "json"}
+
+	defaultSections = []string{
+		IstioOperatorCRSection,
+		ControlPlaneSection,
+		GatewaysSection,
+	}
+
+	verboseSections = []string{
+		IstioOperatorCRSection,
+		ControlPlaneSection,
+		GatewaysSection,
+		PodsSection,
+	}
+)
+
+var (
+	istioOperatorGVR = apimachinery_schema.GroupVersionResource{
+		Group:    iopv1alpha1.SchemeGroupVersion.Group,
+		Version:  iopv1alpha1.SchemeGroupVersion.Version,
+		Resource: "istiooperators",
+	}
+
+	revArgs = revisionArgs{}
+)
+
+func revisionCommand() *cobra.Command {
+	revisionCmd := &cobra.Command{
+		Use:     "revision",
+		Short:   "Revision centric view of Istio deployment",
+		Aliases: []string{"rev"},
+	}
+	revisionCmd.PersistentFlags().StringVarP(&revArgs.manifestsPath, "manifests", "d", "", mesh.ManifestsFlagHelpStr)
+	revisionCmd.PersistentFlags().BoolVarP(&revArgs.verbose, "verbose", "v", false, "print customizations")
+
+	revisionCmd.AddCommand(revisionListCommand())
+	revisionCmd.AddCommand(revisionDescribeCommand())
+	return revisionCmd
+}
+
+// TODO(su225): Fix example, description - command documentation part
+func revisionDescribeCommand() *cobra.Command {
+	describeCmd := &cobra.Command{
+		Use:     "describe",
+		Example: `    istioctl experimental revision describe <revision>`,
+		Short:   "Show details of a revision - customizations, number of pods pointing to it, istiod, gateways etc",
+		Args: func(cmd *cobra.Command, args []string) error {
+			if len(args) == 0 {
+				return fmt.Errorf("revision is not specified")
+			}
+			if len(args) > 1 {
+				return fmt.Errorf("exactly 1 revision should be specified")
+			}
+			if revArgs.verbose && len(revArgs.sections) != 0 {
+				return fmt.Errorf("sections and verbose flags should not be specified at the same time")
+			}
+			isValidFormat := false
+			for _, f := range validFormats {
+				if revArgs.output == f {
+					isValidFormat = true
+					break
+				}
+			}
+			if !isValidFormat {
+				return fmt.Errorf("unknown format %s. It should be %#v", revArgs.output, validFormats)
+			}
+			return nil
+		},
+		RunE: func(cmd *cobra.Command, args []string) error {
+			revArgs.name = args[0]
+			if revArgs.name != "<default>" {
+				if errs := validation.IsDNS1123Label(revArgs.name); len(errs) > 0 {
```

Comment:
Will be good to perform validation in `PreRunE`. Ref [here](https://github.com/istio/istio/blob/96f0b401cdc7794817187334bcbee44ff1193d48/operator/cmd/mesh/install.go#L114).